src/main/java/net/glowstone/net/handler/play/inv/VehicleMoveHandler.java

```diff
@@ -1,12 +1,119 @@
 import net.glowstone.net.message.play.entity.VehicleMoveMessage;
+import net.glowstone.util.Position;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.Statistic;
+import org.bukkit.entity.Boat;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Minecart;
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
+import org.bukkit.event.vehicle.VehicleMoveEvent;
+import org.bukkit.util.NumberConversions;
+import org.bukkit.util.Vector;
 
 public class VehicleMoveHandler implements MessageHandler<GlowSession, VehicleMoveMessage> {
     @Override
     public void handle(GlowSession session, VehicleMoveMessage message) {
-        //TODO: Sent when a player moves in a vehicle. Fields are the same as in Player Position And Look. Note that all fields use absolute positioning and do not allow for relative positioning.
+        GlowPlayer player = session.getPlayer();
+
+        if (!player.isInsideVehicle() || !(player.getVehicle() instanceof Vehicle)) {
+            return;
+        }
+
+        GlowEntity vehicle = (GlowEntity) player.getVehicle();
+
+        Location oldLocation = vehicle.getLocation();
+        Location newLocation = oldLocation.clone();
+        message.update(newLocation);
+
+        // don't let players reach an illegal position
+        if (Math.abs(newLocation.getBlockX()) > 32000000 || Math.abs(newLocation.getBlockZ()) > 32000000) {
+            session.getPlayer().kickPlayer("Illegal position");
+            return;
+        }
+
+        /*
+          don't let players move more than 100 blocks in a single packet
+          if they move greater than 10 blocks, but less than 100, just warn
+          this is NOT robust hack prevention - only to prevent client
+          confusion about where its actual location is (e.g. during login)
+        */
+        if (Position.hasMoved(oldLocation, newLocation) && !player.isDead() && !vehicle.isDead()) {
+            double distance = newLocation.distanceSquared(oldLocation);
+            if (distance > 100 * 100) {
+                session.getPlayer().kickPlayer("You moved too quickly :( (Hacking?)");
+                return;
+            } else if (distance > 100) {
+                GlowServer.logger.warning(session.getPlayer().getName() + " moved too quickly!");
+            }
+        }
+
+        if (!isValidMovement(vehicle, oldLocation, newLocation)) {
+            vehicle.teleport(oldLocation);
+            return;
+        }
+
+        // call move event if movement actually occurred and there are handlers registered
+        if (!oldLocation.equals(newLocation) && VehicleMoveEvent.getHandlerList().getRegisteredListeners().length > 0) {
+            VehicleMoveEvent event = EventFactory.callEvent(new VehicleMoveEvent((Vehicle) vehicle, oldLocation, newLocation));
+
+            if (!event.getTo().equals(newLocation)) {
+                vehicle.teleport(event.getTo(), TeleportCause.PLUGIN);
```

Comment:
Changed the implementation to be more in line with papers. Changing the to or from location of this event does not seem to do anything there.